Python/eof.py

```diff
@@ -0,0 +1,198 @@
+import numpy as np
+from sklearn.decomposition import pca 
+from sklearn import linear_model
+
+from operators import *
+
+def ensemble_reshape(data_cube):
+    """Reshapes data cube into array shape 
+    for eof calculations. 
+    --------------------------------
+    Args: 
+        data_cube (xarray Dataset): datacube 
+            containing an ensemble of 
+            CanRCM4 models 
+    Returns:
+        field_ens (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    """
+    check_keys(data_cube)
+
+    x_field = data_cube['dv']
+    n, p = cell_count(data_cube)
+    field_ens = np.reshape(x_field.values, (p, n))
+
+    return field_ens
+
+def get_land(ensemble_arr):
+    """Gets indeces of nan values to 
+    identify where land or ocean is. 
+    --------------------------------
+    Args: 
+        ensemble_arr (numpy.ndarray): reshaped 2-d array
+            into number of cells x ensemble size
+    Returns:
+        (numpy.ndarray): 2-d array containing
+            the indeces of nan values
+    """
+    return np.isnan(ensemble_arr)
+
+def rand_sample_index(p, frac):
+    """Randomly sample a range of integers. 
+    --------------------------------
+    Args: 
+        p (int): maximum integer to sample
+        frac (float): fractional size of p 
+            to sample
+    Returns:
+        (numpy.ndarray): array containing indeces from 
+            a fraction of 0 to p
+    """
+    index = np.random.choice(np.arange(p), int(frac*p))
+    return index
+
+def pseudo_obs(data_cube, frac=0.02):
+    """Randomly sample a data cube to generate
+    pseudo observations from ensemble. 
+    --------------------------------
+    Args:  
+        data_cube (xarray Dataset): datacube 
+            containing an ensemble of 
+            CanRCM4 models 
+    Returns:
+        y_sample (numpy.ndarray): fraction
+            of random spatially sampled design
+            value grid cells
+        index (numpy.ndarray): index locations 
+            of y_sample from original ensemble size
+    """
+    Y_ens = ensemble_reshape(data_cube)
+    land_idx = get_land(Y_ens)
+    # select only land
+    Y_land = Y_ens[~land_idx.any(axis=1)]
+
+    p, n = Y_land.shape
+
+    # randomly select an ensemble member
+    # to spatially sample
+    i = np.random.randint(0, n-1)
+    y_land = Y_land[:, i]
+
+    # select random indeces
+    index = rand_sample_index(p, frac)
+    y_sample = y_land[index]
+
+    return y_sample, index
+
+def ensemble_to_eof(ensemble_arr):
+    """Perform EOF/PCA dimensionality reduction
+    on ensemble array 
+    --------------------------------
+    Args:  
+        ensemble_arr (numpy.ndarray): reshaped 2-d array
```

Comment:
Do you mean that the arg should be of dimensions num cells by ensemble size? I wouldn't say "reshaped" here, since that necessarily includes a larger context of other processing. Maybe that feels a bit picky, but the whole point of functions is to limit the context that you need to think about to understand them. Everything should be self contained without have to worry about anything else. So just say: "2-day array of size <number of cells> x <number of ensembles>".